src/hazmat/float.rs

```diff
@@ -113,12 +116,68 @@ pub(crate) const fn floor_sqrt(x: u32) -> u32 {
     (low + high) / 2
 }
 
+// Calculate the natural logarithm of a big integer using the relation ln(x) = log₂(x) / log₂(e).
+// Uses fixed-point arithmetic for large values of x (> 2^53).
+#[inline]
+pub(crate) fn ln<const LIMBS: usize>(x: &Uint<LIMBS>) -> f64 {
+    if x <= &Uint::ONE {
+        return 0.0;
+    }
+    let ilog2_x = x.bits_vartime().saturating_sub(1);
+    // if x is small enough to be cast losslessly to an f64 we use the normal 64-bit log2() from `libm`.
+    if ilog2_x < f64::MANTISSA_DIGITS {
```

Comment:
I am not 100% sure, but it seems that 32-bit targets in Rust actually support `f64`, so we may not need to do anything special for them?

Reply:
Oh they do (internally the implementation differs of course). Are you referring to this line in particular? 32-bit targets definitely have access to the `f64` constants.

Reply:
What I meant is whether we need to branch on the target being 32-bit or 64-bit, or just work with `f64` regardless.

Reply:
So when the bits of `x` would fit in a u64, we just use `libm`'s log2 function and that works fine for both 32 and 64 bit targets (see line 146), but `x` is a `Uint<LIMBS>`, and *does* have a different representation on 32 bit targets, so that's what is going on on line 130 - 143: converting `x` to a `u64`. Once it is of that type, 64 bit floating point maths works the same on any target.